DOC: Add developer guidelines for implementing the nan_policy parameter. (#12069)
scipy/scipy: doc/source/dev/api-dev/nan_policy.rst

  function that accepts a single array (and ignoring the possible use of
  ``axis`` for the moment)::

      func([1.0, 3.0, np.nan, 5.0], nan_policy='omit')

  should behave the same as::

      func([1.0, 3.0, 5.0])

  More generally, ``func(a, nan_policy='omit')`` should behave the same as
  ``func(a[~np.isnan(a)])``.

  Unit tests for this property should be used to test functions that
  handle ``nan_policy``.

  For functions that accept two or more arguments but whose values are
  not related, the same idea applies to each input array.  So::

      func(a, b, nan_policy='omit')

  should behave the same as::

      func(a[~np.isnan(a)], b[~np.isnan(b)])

  For inputs with *related* values, the recommended behavior is to omit
  all the values for which any of the related values are ``nan``.  For
  a function with two related array inputs, this means::

      y = func(a, b, nan_policy='omit')

  should behave the same as::

      hasnan = np.isnan(a) | np.isnan(b)  # Union of the isnan masks.
      y = func(a[~hasnan], b[~hasnan])

  The docstring for such a function should clearly state this behavior.

* ``nan_policy='raise'``:

>   Raise a ``ValueError``.

Author: (Getting back to this after a long delay...)

I see that as an implementation detail, and not really part of the API specification.  We might rewrite,  rename, or relocate that function in the future, and that should not change the API spec.  And for some functions, we might not even use that function.  Also, I think of this document as useful for both developers and users, and we don't need to subject the users to those implementation details here.

Having said that, it might be useful to add a section (in this doc, or perhaps a separate one) about the current recommended best practices for implementing the API, with the understanding that the implementation details are subject to change.  To minimize the additional review needed for this PR, I think we can defer that to a later PR.

>   Raise a ``ValueError``.

Reviewer: >Also, I think of this document as useful for both developers and users, and we don't need to subject the users to those implementation details here.

This makes more sense, I was reading this as guidance for those of us maintaining/implementing functions that support `nan_policy` and not necessarily for users.  

> Having said that, it might be useful to add a section (in this doc, or perhaps a separate one) about the current recommended best practices for implementing the API, with the understanding that the implementation details are subject to change. To minimize the additional review needed for this PR, I think we can defer that to a later PR.

Agreed, for several reasons, let's avoid implementation details here.

* ``nan_policy='propagate'``:
  Progagate the ``nan`` value to the output.  This sounds like a
  straightforward specification, but occasionally there are subtleties

>   to be considered--see below.

Author: There are others as well.  The issue reported in https://github.com/scipy/scipy/issues/7818 is not related to handling an `axis` argument.  But see my longer comment below related to this...

>   to be considered--see below.

Reviewer: Just a comment, this seems fairly complicated and likely to require some case by case judgment calls.

collection of 1-d arrays, with the ``axis`` parameter specifying the axis
along which the reduction is to be applied.  If, say::

    func([1, 3, 4])     -> 10.0
    func([2, -3, 8, 2]) ->  4.2
    func([7, 8])        ->  9.5
    func([])            -> -inf

then::

    func([[  1, nan,   3,   4],
          [  2,  -3,   8,   2],
          [nan,   7, nan,   8],
          [nan, nan, nan, nan]], nan_policy='omit', axis=-1)

must give the result::

    np.array([10.0, 4.2, 9.5, -inf])

Edge cases
----------
A function that implements the ``nan_policy`` parameter should gracefully
handle the case where *all* the values in the input array(s) are ``nan``.
The basic principle described above still applies::

    func([nan, nan, nan], nan_policy='omit')

should behave the same as::

    func([])

In practice, when adding ``nan_policy`` to an existing function, it is
not unusual to find that the function doesn't already handle this case
in a well-defined manner, and some thought and design may have to be
applied to ensure that it works.  The correct behavior (whether that be
to return ``nan``, return some other value, raise an exception, or something

> else) will be determined case-by-case.

Author: Done.

  as expected, e.g. ``3 < inf`` is True.
* For the most part, arithmetic works "as expected" with ``inf``,
  e.g. ``inf + inf = inf``, ``-2*inf = -inf``, ``1/inf = 0``,
  etc.
* Many existing functions work "as expected" with ``inf``:
  ``np.log(inf) = inf``, ``np.exp(-inf) = 0``,
  ``np.array([1.0, -1.0, np.inf]).min() = -1.0``, etc.

So while ``nan`` almost always means "something went wrong" or "something
is missing", ``inf`` can in many cases be treated as a useful floating
point value.

It is also consistent with the NumPy ``nan`` functions to not ignore
``inf``::

    >>> np.nanmax([1, 2, 3, np.inf, np.nan])
    inf
    >>> np.nansum([1, 2, 3, np.inf, np.nan])
    inf
    >>> np.nanmean([8, -np.inf, 9, 1, np.nan])
    -inf

Subtleties of ``nan_policy='propagate'``
----------------------------------------
In many case, ``nan_policy='propagate'`` is trivial to implement: just
execute the code in the function without checking for ``nan``, and let the
chips fall where they may.  There are two potential problems with this:

* It may result in warnings that, for the user, are unexpected. E.g. "Why
  are you warning me about invalid values in double scalars when I have

>   already told you to propagate the ``nan`` values?"

Author: You're right, this should provide clearer guidance.  Or maybe no guidance--perhaps my attempt to discuss the subtleties in this document is premature and can be removed.

The motivating example--indeed, the only example at the moment--is the issue reported in https://github.com/scipy/scipy/issues/7818.  The uniform filter is just a sliding windowed mean of the data.  There are at least three ways such a calculation could be implemented, and each one propagates `nan` differently.  In the "naive" implementation, the `nan`s in the output are limited to a short section with length equal to the window length.  For the implementation in `ndimage.uniform_filter`, the `nan` in the output propagates from the first occurrence of the input `nan` to the end of the array.  If the filter was calculated using the FFT, all the output values would be `nan`.  Mathematically (and in the absence of `nan`), all three implementations are equivalent--and a user shouldn't have to care about how that calculation is actually implemented.  But when we consider propagating a `nan`, the implementation details can become part of the API.  For the case of the uniform filter, do we want to promise that `nan` will propagate according to the naive implementation (as the user in that issue expected)?  This might mean checking for `nan` in advance, or checking for it during the computation, and switching to the naive implementation if a `nan` is encountered.  Or do we not bother trying to promise exactly how `nan`s will propagate, and instead state that the behavior should not be relied on because it is implementation-dependent and might change if the underlying implementation changes?  I'm not sure at the moment, which is why I think maybe I should take this out of the document.

Regarding warnings:  I've been trying to cook up an example or two where I can say "Here's a naive implementation of the 'propagate' behavior for a function that generates an undesirable warning", but I haven't had much luck.  All my simple examples fail to generate a warning, and I don't recall if I originally had an example that motivated me to write this.  So yet another reason to perhaps just take this section out.

>   already told you to propagate the ``nan`` values?"

Reviewer: > You're right, this should provide clearer guidance. Or maybe no guidance--perhaps my attempt to discuss the subtleties in this document is premature and can be removed.

 > The motivating example
> ... [eliding full quoting of a long section]
> I'm not sure at the moment, which is why I think maybe I should take this out of the document.

I'm in agreement on taking this part out of the document, if examples come up later we can always add this language, or some updated version of it, into the document.
 
> Regarding warnings: I've been trying to cook up an example or two where I can say "Here's a naive implementation of the 'propagate' behavior for a function that generates an undesirable warning", but I haven't had much luck. All my simple examples fail to generate a warning, and I don't recall if I originally had an example that motivated me to write this. So yet another reason to perhaps just take this section out.

Hmm, I can't really think of any obvious example either so maybe is best to just take it out for now.

> execute the code in the function without checking for ``nan``, and let the
> chips fall where they may.  There are two potential problems with this:
> 
> * It may result in warnings that, for the user, are unexpected. E.g. "Why
>   are you warning me about invalid values in double scalars when I have
>   already told you to propagate the ``nan`` values?"

Reviewer: ```suggestion
execute the code in the function without checking for ``nan``.
There are potential problems with this approach. 

For example:

```

Reviewer: @WarrenWeckesser I left a suggestion here to cut this part of the section given we both seem to agree it's not very clear and we can always clarify and add later.

* How a ``nan`` propagates will be implementation dependent, and in some
  cases may surprise a user.  A concrete example is reported in GitHub
  issue https://github.com/scipy/scipy/issues/7818, where a user was
  surprised that a ``nan`` propagated all the way to the end of an array,
  rather than being limited to outputs for which a naive implementation

>   would only be affected by inputs within the specified window.

Reviewer: This second part of the section seems clear though.

  Generally, whether ``'propagate'`` needs to be handled as a special
  case will be decided on a case-by-case basis.

How *not* to implement ``nan_policy``
-------------------------------------
In the past (and possibly currently), some ``stats`` functions handled
``nan_policy`` by using a masked array to mask the ``nan`` values, and
then computing the result using the functions in the ``mstats`` subpackage.
The problem with this approach is that the masked array code might convert
``inf`` to a masked value, which we don't want to do (see above).  It also

> means that, if care is not taken, the return value will be a masked array,

Author: That's my preference: masked arrays should not be generated if none of the inputs are masked arrays.  I think of that as a *general* guideline, and not just a guideline for `nan_policy` handling.

In general, there is nothing wrong with propagating a masked array if one or more inputs is a masked array.  But what I've written here is only for the case where the inputs are not masked arrays.  We shouldn't generate masked arrays in that case.

> means that, if care is not taken, the return value will be a masked array,

Reviewer: > That's my preference: masked arrays should not be generated if none of the inputs are masked arrays. I think of that as a _general_ guideline, and not just a guideline for `nan_policy` handling.

This makes sense.
 
> In general, there is nothing wrong with propagating a masked array if one or more inputs is a masked array. But what I've written here is only for the case where the inputs are not masked arrays. We shouldn't generate masked arrays in that case.

Also makes sense. 

I guess we should leave this section as is, it seems a bit clearer to me on the second read through. Marking this resolved.
